akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java

```diff
@@ -0,0 +1,129 @@
+package jdocs.persistence.query;
+
+import akka.actor.ActorSystem;
+import akka.japi.Pair;
+import akka.persistence.PersistentRepr;
+import akka.persistence.query.EventEnvelope;
+import akka.persistence.query.Offset;
+import akka.serialization.Serialization;
+import akka.serialization.SerializationExtension;
+import akka.stream.*;
+import akka.stream.stage.GraphStage;
+import akka.stream.stage.GraphStageLogic;
+import akka.stream.stage.OutHandler;
+import akka.stream.stage.TimerGraphStageLogic;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+
+// #events-by-tag-publisher
+public class MyEventsByTagSource extends GraphStage<SourceShape<EventEnvelope>> {
+   private Outlet<EventEnvelope> out = Outlet.create("MyEventByTagSource.out");
+   private static final String CONTINUE = "CONTINUE";
+   private static final int LIMIT = 1000;
+   private final Connection connection;
+   private final String tag;
+   private final long initialOffset;
+   private final Duration refreshInterval;
+
+   public MyEventsByTagSource(Connection connection, String tag, long initialOffset, Duration refreshInterval) {
+      this.connection = connection;
+      this.tag = tag;
+      this.initialOffset = initialOffset;
+      this.refreshInterval = refreshInterval;
+   }
+
+   @Override
+   public SourceShape<EventEnvelope> shape() {
+      return SourceShape.of(out);
+   }
+
+   class Logic extends TimerGraphStageLogic implements OutHandler {
+
+      private ActorSystem system = ((ActorMaterializer) materializer()).system();
```

Comment:
Not nice with the cast in example code, but `akka.stream.ActorMaterializerHelper#downcast` is internal...

Reply:
Indeed